File: jvm/src/main/java/com/muchq/cards/Card.java
package com.muchq.cards;

import java.util.List;

public record Card(Suit suit, Rank rank, Facing facing) {
    private static final List<Suit> SUITS = List.of(Suit.CLUBS, Suit.DIAMONDS, Suit.HEARTS, Suit.SPADES);
    private static final List<Rank> RANKS = List.of(
            Rank.TWO, Rank.THREE, Rank.FOUR, Rank.FIVE, Rank.SIX, Rank.SEVEN, Rank.EIGHT,
            Rank.NINE, Rank.TEN, Rank.JACK, Rank.QUEEN, Rank.KING, Rank.ACE);
    
    public static Card forIndex(int index) {
        Suit s = SUITS.get(index % 4);
        Rank r = RANKS.get(index % 13);
        return new Card(s, r, Facing.DOWN);
    }
}

File: cpp/cards/card.cc
#include "card.h"

#include <unordered_map>

namespace cards {

const Card Card::flipped() {
  return Card(suit, rank, static_cast<Facing>((static_cast<int>(facing) + 1) % 2));
}

const std::string Card::debug_string() {
  std::string repr = "";
  if (RANKS.find(rank) != RANKS.end()) {
    repr.append(RANKS.at(rank));
  } else {
    repr.append("unknown_rank");
  }
  repr.append("_");
  if (SUITS.find(suit) != SUITS.end()) {
    repr.append(SUITS.at(suit));
  } else {
    repr.append("unknown_suit");
  }

  switch (facing) {
    case Facing::Up:
      repr.append("_Up");
    default:
      repr.append("_Down");
  }

  return repr;
}
}  // namespace cards